internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,393 @@
+func (c *JobContainer) ID() string {
+	return c.id
+}
+
+// Shutdown gracefully shuts down the container.
+func (c *JobContainer) Shutdown(ctx context.Context) error {
+	log.G(ctx).WithField("id", c.id).Debug("shutting down job container")
+
+	return c.shutdown(ctx)
+}
+
+// shutdown will loop through all the pids in the container and send a signal to exit.
+// If there are no processes in the container it will just early return nil.
+// If the all processes exited message is not received within the context timeout set, it will
+// terminate the job.
+func (c *JobContainer) shutdown(ctx context.Context) error {
+	pids, err := c.job.Pids()
+	if err != nil {
+		return errors.Wrap(err, "failed to get pids in container")
+	}
+
+	if len(pids) == 0 {
+		return nil
+	}
+
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			return err
+		}
+	}
+
+	select {
+	case <-c.exited:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+// PropertiesV2 is not implemented for job containers. This is just to satisfy the cow.Container interface.
+func (c *JobContainer) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (*hcsschema.Properties, error) {
+	return nil, errors.New("propertiesV2 call not implemented for job containers")
+}
+
+// Properties is not implemented for job containers. This is just to satisfy the cow.Container interface.
+func (c *JobContainer) Properties(ctx context.Context, types ...schema1.PropertyType) (*schema1.ContainerProperties, error) {
+	return nil, errors.New("properties call not implemented for job containers")
+}
+
+// Terminate terminates the job object (kills every process in the job).
+func (c *JobContainer) Terminate(ctx context.Context) error {
+	log.G(ctx).WithField("id", c.id).Debug("terminating job container")
+
+	if err := c.job.Terminate(1); err != nil {
+		log.G(ctx).WithField("id", c.id).WithError(err).Error("failed to terminate containers job object")
+
+		return err
+	}
+	return nil
+}
+
+// Wait synchronously waits for the container to shutdown or terminate. If
+// the container has already exited returns the previous error (if any).
+func (c *JobContainer) Wait() error {
+	<-c.waitBlock
+	return c.waitError
+}
+
+func (c *JobContainer) waitBackground(ctx context.Context) {
+	// Wait for there to be an init process assigned.
+	<-c.init.initBlock
+
+	// Once the init process finishes, if there's any other processes in the container we need to signal
+	// them to exit.
+	<-c.init.proc.waitBlock
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	if err := c.Shutdown(ctx); err != nil {
+		c.Terminate(ctx)
+	}
+
+	c.closedWaitOnce.Do(func() {
+		c.waitError = c.init.proc.waitError
+		close(c.waitBlock)
+	})
+}
+
+// Polls for notifications from the job objects assigned IO completion port.
+func (c *JobContainer) pollJobMsgs(ctx context.Context) {
+	if c.job == nil {
+		return
+	}
```

Comment:
What scenario could this happen in?

Reply:
Wouldn't happen in our cases as we would call `Create` and a job would always be present but left it in just in case this ever gets reworked to nil the job on shutdown/close, or there's ever some use to just creating a `JobContainer` object standalone etc.